File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
from typing import List
from collections import defaultdict

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefix_sums = defaultdict(int)
        prefix_sums[0] = 1

        count = 0
        total = 0

        for num in nums:
            total += num
            count += prefix_sums[total - k]
            prefix_sums[total] += 1

        return count

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.ts
function subarraySum(nums: number[], k: number): number {
    const prefixSums: Map<number, number> = new Map();
    prefixSums.set(0, 1); // Handle subarrays starting at index 0

    let count = 0;
    let total = 0;

    for (const num of nums) {
        total += num;

        // Check if (total - k) has been seen before
        const matching = prefixSums.get(total - k) || 0;
        count += matching;

        // Update frequency map
        prefixSums.set(total, (prefixSums.get(total) || 0) + 1);
    }

    return count;
}
